### PR TITLE
pianoroll: nudge/snap while dragging notes

### PR DESCRIPTION
**Pull request by rghvdberg**
Adds combobox for switching behavior of the snapping to grid when dragging.
Old behavior (nudge) is default.
Notice that snapping is relative to the dragged note and not the 1st note in the selected range.

It's essentially the same as #4357 but I lost that fork and just started over.
lmms_snap

### CODE REVIEW

**Comment by Spekular** [comment]
Would you consider adding snapping to the end of the note as well? (Choosing whichever is closest to the unquantized position, as with shift drag in the song editor). It would be convenient for reversed samples and such, and shouldn't interfere much with the existing functionality in most situations.

I think this is worth adding regardless, but it's a nice-to-have if you're not opposed.

**Comment by rghvdberg** [comment]
Didn't think of that. Will do.

Op ma 14 dec. 2020 11:25 schreef Spekular <notifications@github.com>:

File: include/PianoRoll.h
Comment (rghvdberg) [reply include/PianoRoll.h]: fixed

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:2641]: Trailing whitespaces:
```suggestion
					else
```

File: src/gui/editors/PianoRoll.cpp
Comment (PhysSong) [inline src/gui/editors/PianoRoll.cpp:4431]: Translations may break this.

**Comment by superpaik** [comment]
Suggestion: to me it's a bit odd that when snap activated notes are snapping both from the start and the end of the note, and make it seem that the movement it's kind of random (though I see it is as intended). Maybe it would be better if snap has in mind the direction of the movement. If going left, snap with the start of the note. If going right, snap with the end of the note.
I don't know it is correct or not, but when in pianoroll inside beat+bass editor, "no duration" notes do not seem to follow the same behavior as "normal" notes.

**Comment by rghvdberg** [comment]
Snapping that depends on the direction of drag seems like a bad idea. What if I just want to move 1/4 note to the right? I'd first have to drag beyond that note and then drag left to finally land there.
We could use a modifier key to enable snap to end and the default being snap to start.
I don't know if there are any modifier keys left to implement in pianoroll though.

**Comment by rghvdberg** [comment]
I only tested (very shortly) in pianoroll in the song editor. I'll look at the bb editor case.

File: src/gui/editors/PianoRoll.cpp
Comment (rghvdberg) [reply src/gui/editors/PianoRoll.cpp:4431]: Would using `m_snapModel.getValue()` be better ?
It returns 0 (Nudge) and 1 (Snap).
I am uneasy using 'magic numbers' in code. Maybe add an `enum` for these values?

**Comment by rghvdberg** [comment]
found a bug :-(
when you drag a note over a second note which the same pitch, the second note is moved a bar ...

https://user-images.githubusercontent.com/10710663/104838584-1ab38280-58bc-11eb-87a6-370b25a23e62.mp4

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:4431]: I'd make an enum and compare to that.

File: include/PianoRoll.h
Comment (IanCaio) [inline include/PianoRoll.h:307]:
```suggestion
	void dragNotes(int x, int y, bool alt, bool shift, bool ctrl);
```

From what I read in the code it seems that `m_draggedNote` is being used with the same goal as `m_currentNote`. Using the latter would probably fix the bug with dragging a note on top of another one (since `m_currentNote` is set on the `mousePressEvent` and will not change during `mouseMoveEvent`) and will prevent adding unnecessary variables to the method and class.

File: src/gui/editors/PianoRoll.cpp
Comment (IanCaio) [reply src/gui/editors/PianoRoll.cpp:4431]: There's already an enum being used for `m_gridMode`, can't we just reuse that? I think just like the enum, the ComboBox model will go from 0 up, then you can cast and compare them.

File: include/PianoRoll.h
Comment (IanCaio) [inline include/PianoRoll.h:262]:
```suggestion
		//gridFree
```
I'd remove or comment this out while it's not implemented

File: src/gui/editors/PianoRoll.cpp
Comment (IanCaio) [inline src/gui/editors/PianoRoll.cpp:2619]:
```suggestion
					// Quantize first note
```

**Review by IanCaio — CHANGES_REQUESTED** [review]
Just an initial review. As said in Discord, I think the bug that you found is being caused by calling `noteUnderMouse` repeatedly to get the note being dragged on the `mouseMoveEvent` method. When two notes are in the same position you might end up with the wrong note on the `dragNotes` call. `m_draggedNote` seems redundant when you have `m_currentNote` and I think you could get rid of it and use the latter instead, fixing the bug and also removing an unnecessary variable.

I wrote a few other comments, but before going further with the review I was thinking about suggesting a different approach on the logic that snaps the note to the grid. Why not using the mouse position, quantizing it and then just calculating the offset necessary to get the note from where it is to that quantized mouse position? It sounds to me like it would make the logic simpler (wouldn't require copying the note, quantizing it and moving it around) and it seems like it would work as intended. The only difference in behavior that I can think of is that the position of the mouse in relation to the grid would matter more than the position of the mouse against the initial position of the note. Would have to check in terms of UX if that behavior is annoying or not, but if it is it shouldn't be too complicated to add the offset against the initial position of the note to the math.

I can help you out writing this alternative approach if you like!

**Comment by superpaik** [comment]
It woks good to me! Tested on regular and "no-duration" (beats) notes.
I think that using the mouse position instead the first note was a great idea! 
Thanks!

**Review by IanCaio — APPROVED** [review]

File: src/gui/editors/PianoRoll.cpp
Comment (Spekular) [inline src/gui/editors/PianoRoll.cpp:2569]: Why not move this so off_ticks is only quantized if you're in nudge mode? Then you should be able to use off_ticks for the snap calculations as well, instead of recalculating it from mouse positions and stuff below.

**Review by Spekular — COMMENTED** [review]
The snapping code looks unnecessarily long to me, and reasonably it seems that the logic for snap vs nudge should be on the same level, rather than snap being calculated "inside" the nudge logic and overriding it.

**Comment by IanCaio** [comment]
> The snapping code looks unnecessarily long to me, and reasonably it seems that the logic for snap vs nudge should be on the same level, rather than snap being calculated "inside" the nudge logic and overriding it.

Ideally the whole `dragNotes` method could use some refactoring that would make it easier to have nudge and snap more explicitly separated. I'm not sure how I would do it without rewriting a lot of the method. Since it doesn't make it worse, I'd go with that approach for now and leave that method overhaul for when we are out of that tight refactoring schedule.

File: src/gui/editors/PianoRoll.cpp
Comment (IanCaio) [reply src/gui/editors/PianoRoll.cpp:2569]: I'm not sure we can really reuse `off_ticks` for the snap mode: `off_ticks` is the offset from the initial position the mouse clicked in ticks, while the position we calculate in the snap mode is the absolute mouse position in ticks on the time line. Maybe if we added `m_moveStartX * TimePos::ticksPerBar() / m_ppb;` to it, but then we are having to do the conversion to ticks again anyways, right?

Comment (Spekular) [reply src/gui/editors/PianoRoll.cpp:2569]: `note->oldPos().getTicks() + off_ticks` should give the absolute position, no?

Comment (Spekular) [reply src/gui/editors/PianoRoll.cpp:2569]: In fact, this is exactly what the existing code already does at line 2598 (2645 in the PR version at time of writing).

```
int pos_ticks = note->oldPos().getTicks() + off_ticks;
...
note->setPos(TimePos(pos_ticks));
```

Comment (IanCaio) [reply src/gui/editors/PianoRoll.cpp:2569]: Sure, I'm giving it a shot. Though I still think this will become a rabbit hole and I'll end up rewriting more than expected.

Comment (IanCaio) [reply src/gui/editors/PianoRoll.cpp]: Last two commits separate the nudge and snap logic more explicitly and reuse off_ticks as you suggested. It required some rearrangement of code. Method still needs refactoring and I tried to rewrite the least possible. A bug that is present on `master` was fixed on the calculation of the boundaries limits for the note keys.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [inline src/gui/editors/PianoRoll.cpp:1197]: Since we repeat this function call a few times, can we at least fix this? Just pass the event or modifiers in and test in the function for what we need.

Comment (Spekular) [reply src/gui/editors/PianoRoll.cpp:1197]: I'm not strongly for or against this change, but wouldn't that couple dragNotes more closely to (1) the Qt UI and (2) these specific modifiers? If for example the argument "alt" were renamed to "unquantized" (or inverted and renamed to "quantized") then the dragNotes function is independent of which shortcut is used to perform unquantized actions.

**Review by Spekular — APPROVED** [review]
LGTM after Ian's latest changes.

File: src/gui/editors/PianoRoll.cpp
Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:1197]: Dragging notes is inherently a UI thing, not a core thing.

I do like the idea of changing the names from `ctrl` `alt` and `shift`, and I would rather just pass in a `modifiers` value to the function.

Comment (Spekular) [reply src/gui/editors/PianoRoll.cpp:1197]: > Dragging notes is inherently a UI thing, not a core thing.

Sure, but at the moment this function contains a lot of logic that doesn't rely on the UI. The same quantization strategies are relevant regardless of how the move is input.

Comment (IanCaio) [reply src/gui/editors/PianoRoll.cpp:1197]: I think the renaming of the arguments and possibly compressing them in a single one seems to be the middle ground: We keep the method decoupled from the Qt event and make the modifiers more generic (allowing us to change them later).

Can I suggest that we do it separately though? As I mentioned on a previous comment, this method needs a bit of an overhaul. Just renaming the arguments would also require some thought because some modifiers are used for multiple actions depending on context. I think both the method refactoring and changing of arguments could be done in a single separate PR, they would both benefit on each other.

Comment (Veratil) [reply src/gui/editors/PianoRoll.cpp:1197]: > Can I suggest that we do it separately though?

Sure. 👍

**Comment by IanCaio** [comment]
Merging tomorrow if there are no objections. There's one conflict but it can be quickly fixed in the web editor before merging.

**Pull request merged**
